### PR TITLE
fix: infinite loop on tls sock when BIO is full and client renegotiates the handshake

### PR DESCRIPTION
**Pull request by kostasrim**
* fixes the infinite loop cause by this corner case

I am also adding a test case in df repo for this :)

### CODE REVIEW

**Comment by kostasrim** [comment]
@romange I sincerely ask for the commit to include @chakaz as well! It was a tag team effort :rocket: :partying_face:

File: io/io.cc [hunk 1]
```diff
@@ -137,7 +137,6 @@ Result<size_t> PrefixSource::ReadSome(const iovec* v, uint32_t len) {
 error_code Sink::Write(const iovec* v, uint32_t len) {
   return ApplyExactly(v, len,
                       [this](const auto* v, uint32_t len) { return WriteSomeBytes(v, len, this); });
-  return error_code{};
```
Comment (chakaz) [inline io/io.cc:140]: Did you intentionally remove this?

**Review by chakaz — APPROVED** [review]

**Review by romange — COMMENTED** [review]
Is it possible to cover with.tests?

File: util/tls/tls_socket.cc [hunk 2]
```diff
@@ -326,10 +326,23 @@ SSL* TlsSocket::ssl_handle() {
 }
 
 auto TlsSocket::MaybeSendOutput() -> error_code {
-  // this function is present in both read and write paths.
+  // This function is present in both read and write paths.
   // meaning that both of them can be called concurrently from differrent fibers and then
   // race over flushing the output buffer. We use state_ to prevent that.
   if (state_ & WRITE_IN_PROGRESS) {
+    // Do not remove the Yield from here because it can enter an infinite loop
```
Comment (romange) [inline util/tls/tls_socket.cc:333]: seems that a similar thing can happen with `TlsSocket::HandleRead()`

File: util/tls/tls_socket.cc [hunk 3]
```diff
@@ -326,10 +326,23 @@ SSL* TlsSocket::ssl_handle() {
...
+    // which can be described by the following chore:
+    // Fiber1 (Connection fiber):
+    // Reads SSL renegotiation request
+    // Sets WRITE_IN_PROGRESS in state_
+    // Tries to write renegotiation response to socket
+    // BIO (SSL buffer) gets filled up
+    // Fiber2 (runs command):
+    // Tries to write response to socket
+    // Can't write to socket because WRITE_IN_PROGRESS is set
+    // Live blocks indefinitely
```
Comment (romange) [inline util/tls/tls_socket.cc:343]: Most probably it fixes the issue but we do not have much visibility to these spinning behaviors inside `TlsSocket::RecvMsg` and `TlsSocket::SendBuffer`.
it seems we can identify them. For example, count number of consecutive loops that `op_val == 0` inside  SendBuffer and if it crosses some hardcoded limit, issue an warning.  maybe you can think of something more elegant. in any case I want to see more logs around these loops

File: util/tls/tls_socket.cc [hunk 4]
```diff
@@ -165,6 +165,22 @@ auto TlsSocket::Close() -> error_code {
   return next_sock_->Close();
 }
 
+class SpinCounter {
+ public:
+  explicit SpinCounter(size_t limit) : limit_(limit) {
```
Comment (kostasrim) [inline util/tls/tls_socket.cc:170]: @romange The way I would properly implement this is by mimicking the interface of the type I am trying to collect stats about. So in our case that's an int, so I would introduce a wrapper class, overload the operators and I would collect the stats internally such that the variable itself can track the changes and do the printing (because now we need `SpinCounter.Check()` on a separate variable) but this seemed to be an overkill for this so I opted in for the simplest and smallest snippet/class utlity possible

File: util/tls/tls_socket.cc [hunk 5]
```diff
@@ -176,6 +192,7 @@ io::Result<size_t> TlsSocket::RecvMsg(const msghdr& msg, int flags) {
 
   Engine::MutableBuffer dest{reinterpret_cast<uint8_t*>(io->iov_base), io->iov_len};
   size_t read_total = 0;
+  SpinCounter spin_count(20);
```
Comment (kostasrim) [inline util/tls/tls_socket.cc:195]: `20` is adhoc, looking for suggestions for the heuristic

File: util/tls/tls_socket.cc (hunk 4 above)
Comment (kostasrim) [reply util/tls/tls_socket.cc:170]: Also if we need it elsewhere, I am happy to generalize move it somewhere else :)

File: util/tls/tls_socket.cc [hunk 6]
```diff
@@ -165,6 +165,24 @@ auto TlsSocket::Close() -> error_code {
...
 
+class SpinCounter {
+ public:
+  explicit SpinCounter(size_t limit) : limit_(limit) {
+  }
+
+  void Check(bool condition) {
+    current_it_ = condition ? current_it_ + 1 : 0;
+    if (current_it_ >= limit_) {
+      LOG(WARNING) << "IO loop spin limit reached: " << limit_;
```
Comment (romange) [inline util/tls/tls_socket.cc:176]: this warning will appear at line 176 instead of appearing in the loop of the function that is deadlocked.

File: util/tls/tls_socket.cc (hunk 5 above)
Comment (romange) [reply util/tls/tls_socket.cc:197]: it's fine

File: util/tls/tls_socket.cc (hunk 6 above)
Comment (kostasrim) [reply util/tls/tls_socket.cc:176]: GOOOD catch! Will address this

File: util/tls/tls_socket.cc [hunk 7]
```diff
@@ -165,6 +165,34 @@ auto TlsSocket::Close() -> error_code {
...
+    return limit_;
+  }
+
+ private:
+  const size_t limit_;
+  size_t current_it_{0};
+};
+
+#define LOG_IF_SPIN_COUNT_LIMIT_REACHED(spin_count, res)                  \
+  if (spin_count.Check(res == 0)) {                                       \
```
Comment (kostasrim) [inline util/tls/tls_socket.cc:192]: I didn't bother adding a `cond` instead of `res` since we only compare against `0` on all use cases so I didn't bother making it more generic nor did I bother to wrap it in `do/while(0)`

**Review by romange — DISMISSED** [review]
Thanks!

**Comment by romange** [comment]
@kostasrim  can you please sign your commits?

File: util/tls/tls_socket.cc (hunk 7 above)
Comment (chakaz) [reply util/tls/tls_socket.cc:192]: Can you use `LOG_IF()` instead?

File: util/tls/tls_socket.cc [hunk 8]
```diff
@@ -165,6 +165,33 @@ auto TlsSocket::Close() -> error_code {
...
+    return limit_;
+  }
+
+ private:
+  const size_t limit_;
+  size_t current_it_{0};
+};
+
+#define LOG_IF_SPIN_COUNT_LIMIT_REACHED(spin_count, res) \
+  LOG_IF(WARNING, spin_count.Check(res == 0))            \
```
Comment (chakaz) [inline util/tls/tls_socket.cc:192]: I meant to use `LOG_IF` directly instead of creating `LOG_IF_SPIN_COUNT_LIMIT_REACHED`, but your call :)

**Review by chakaz — APPROVED** [review]

**Review by chakaz — DISMISSED** [review]

File: util/tls/tls_socket.cc [hunk 9]
```diff
@@ -290,6 +317,8 @@ io::Result<size_t> TlsSocket::SendBuffer(Engine::Buffer buf) {
     }
 
     int op_val = *op_result;
+    LOG_IF(WARNING, spin_count.Check(op_val == 0))
```
Comment (adiholden) [inline util/tls/tls_socket.cc:320]: This can lead potentially to logs getting full right?

Comment (kostasrim) [reply util/tls/tls_socket.cc:320]: No, look above what `Check` is doing internally. It resets the counter, so it acts as `LOG_EVERY_N`

Comment (adiholden) [reply util/tls/tls_socket.cc:320]: In this case N is 20 so if we have a client that is not reading from the socket for some reason we will write every 20 yields

Comment (kostasrim) [reply util/tls/tls_socket.cc:320]: yes that's true. I wrote in a previous comment this is a heuristic, we can always adjust :shrug:

Comment (adiholden) [reply util/tls/tls_socket.cc:320]: I would either use log every T here or would change the spin count to grow exponentialy i,e first time print after x times next after x^2 times etc

Comment (chakaz) [reply util/tls/tls_socket.cc:320]: we can also use LOG_EVERY_N_SEC() with sec=600 for example..

Comment (adiholden) [reply util/tls/tls_socket.cc:320]: we dont use absl logging - LOG_EVERY_N_SEC
glog have LOG_EVERY_T which is the same

**Review by romange — APPROVED** [review]
